### PR TITLE
[FLINK-32570][runtime][stream] Deprecates @Public/@PublicEvolving API that uses Flink's Time classes

### PR DESCRIPTION
**Pull request by XComp**
## What is the purpose of the change

A preparation for FLINK-14068 to finally remove `org.apache.flink.api.common.time.Time` and `org.apache.flink.streaming.api.windowing.time.Time` (related to [FLIP-335](https://cwiki.apache.org/confluence/display/FLINK/FLIP-335%3A+Removing+Flink%27s+Time+classes)).

## Brief change log

I went through any class that was listed by the following bash command:
```
grep "import.*org\.apache\.flink\..*\.time.Time;" $(find . -name "*.java" -exec grep -H "@Public\|@Experimental" {} \; | cut -d':' -f1 | sort -u) | cut -d: -f1
```
and looked for occurrences of the regex `Time[^a-zA-Z0-9]` in the code and docs.

The JavaDoc was updated accordingly pointing to the replacing method.

## Verifying this change

Manual check - I appreciate any proposal on how to verify the change in a better way. :thinking: 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: yes
 
## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by zentol** [comment]
Has there been a FLIP yet for the removal of the Time class from user-facing APIs?

**Comment by XComp** [comment]
I assumed the [discussion mentioned in FLINK-14068](https://lists.apache.org/thread/76yywnwf3lk8qn4dby0vz7yoqx7f7pkj) and FLINK-14068 itself are good enough as a conclusion. But I can go ahead and formalize the steps in a FLIP in addition to that.

**Comment by zentol** [comment]
> I assumed the [discussion mentioned in FLINK-14068](https://lists.apache.org/thread/76yywnwf3lk8qn4dby0vz7yoqx7f7pkj) and [FLINK-14068](https://issues.apache.org/jira/browse/FLINK-14068) itself are good enough as a conclusion.

Which is a fair point, but I' just not sure if a discussion thread from 4 years ago where 2.0 wasn't something tangible meets our current bar for making breaking changes :thinking:

**Comment by XComp** [comment]
Yup, I agree. I feel like it makes sense to summarize everything in a FLIP. I created [FLIP-335](https://cwiki.apache.org/confluence/display/FLINK/FLIP-335%3A+removing+flink's+time+classes) and started a [discussion thread](https://lists.apache.org/thread/48ysrg1rrtl8s1twg9wmx35l201hnc2w).

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateTtlConfig.java [hunk 1]
```diff
@@ -53,13 +54,14 @@ public class StateTtlConfig implements Serializable {
...
     public static final StateTtlConfig DISABLED =
-            newBuilder(Time.milliseconds(Long.MAX_VALUE))
+            newBuilder(Duration.ofMillis(Long.MAX_VALUE))
                     .setUpdateType(UpdateType.Disabled)
                     .build();
 
     /**
      * This option value configures when to update last access timestamp which prolongs state TTL.
      */
+    @PublicEvolving
```
Comment (1996fanrui) [inline flink-core/src/main/java/org/apache/flink/api/common/state/StateTtlConfig.java:64]: nit: The public class `StateTtlConfig` is `PublicEvolving`, do we need to mark all internal class as well?

IIUC, all public internal class or fields are `PublicEvolving` if public class is `PublicEvolving`, right?

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateTtlConfig.java [hunk 2]
```diff
@@ -121,6 +125,10 @@ public StateVisibility getStateVisibility() {
 
     @Nonnull
```
Comment (1996fanrui) [inline flink-core/src/main/java/org/apache/flink/api/common/state/StateTtlConfig.java:126]:
```suggestion
    /** @deprecated Use {@link #getTimeToLive()} */
    @Deprecated
    @Nonnull
```

File: flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java [hunk 3]
```diff
@@ -72,9 +72,23 @@ public static RestartStrategyConfiguration fixedDelayRestart(
      * @param restartAttempts Number of restart attempts for the FixedDelayRestartStrategy
      * @param delayInterval Delay in-between restart attempts for the FixedDelayRestartStrategy
      * @return FixedDelayRestartStrategy
+     * @deprecated Use {@link #fixedDelayRestart(int, Duration)}
      */
+    @Deprecated
```
Comment (1996fanrui) [inline flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java:77]: IIUC, we don't need to change this class. 

The whole `RestartStrategies` class has been marked as `Deprecated` in the master branch.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 4]
```diff
@@ -810,7 +811,8 @@ public RestartStrategies.RestartStrategyConfiguration getRestartStrategy() {
      * @param numberOfExecutionRetries The number of times the system will try to re-execute failed
      *     tasks.
      * @deprecated This method will be replaced by {@link #setRestartStrategy}. The {@link
-     *     RestartStrategies#fixedDelayRestart(int, Time)} contains the number of execution retries.
+     *     RestartStrategies#fixedDelayRestart(int, Duration)} contains the number of execution
+     *     retries.
```
Comment (1996fanrui) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:815]: Not needed as well

**Review by 1996fanrui — COMMENTED** [review]
Hi @XComp , thanks for the contribution!

Overall looks good to me, I left some comments. Please take a look in your free time, thanks~

Also, this PR needs to rebase master branch.

File: flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java (hunk 3 above)
Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java:77]: That's correct. This change can be reverted. :+1:

Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java:77]: I went through the change once more and came to the conclusion that we could keep this change. You're right when stating that the `RestartStrategies` are going away, anyway. But it doesn't hurt that we do the deprecation "independently" of the `RestartStrategies` efforts. That way, we would still have a consistent code base if (for whatever reason) [FLIP-381](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=278464992) gets reverted. WDYT?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java (hunk 4 above)
Comment (XComp) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:815]: See my reasoning in the [above comment](https://github.com/apache/flink/pull/22984#discussion_r1441914909)

**Comment by XComp** [comment]
Thanks for picking this up. We should merge FLINK-32570 soon'ish to not miss it for 1.19.

I reiterated over the PR once more with the following changes:
* I rebased the branch to most-recent master. 
* I removed the hotfix PRs which added API-related annotations. That is out-of-scope for this PR.
* I improved the naming for some of the methods (`getWindowDuration` becomes `getWindowSize`)
* I reorganized/squashed commits into two (it makes sense to look at the two commits individually when doing the next review)

PTAL

File: flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java (hunk 3 above)
Comment (1996fanrui) [reply flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java]: You are right, this change is useful if FLIP-381 is reverted. But it's very unlikely. And these changes can be done when it happens.

So far, this change isn't required, and I'm fine to have it. So up to you.

**Review by 1996fanrui — APPROVED** [review]
> Thanks for picking this up. We should merge [FLINK-32570](https://issues.apache.org/jira/browse/FLINK-32570) soon'ish to not miss it for 1.19.

Yeah, I found this task from the 2.0 tasks, and this PR is must have in 1.19. Thanks @XComp for the quick update!

LGTM. Feel free to merge it if you are ready.

**Review by 1996fanrui — CHANGES_REQUESTED** [review]
Sorry, the CI fails. Please check and fix it in your free time, thanks~

File: flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java (hunk 3 above)
Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java:95]: Another reason would be that not touching `RestartStrategies` means that we have a (soft) dependencies between FLIP-381 and FLIP-335. With the alternative methods in `RestartStrategies` in place, we can go ahead with FLIP-335 in a more straight-forward fashion and don't need to wait with removing the `Time` classes until  FLIP-381 is resolved. :thinking:

Comment (1996fanrui) [reply flink-core/src/main/java/org/apache/flink/api/common/restartstrategy/RestartStrategies.java:95]: Thanks for the clarifying. LGTM

**Comment by 1996fanrui** [comment]
@flinkbot run azure

**Comment by XComp** [comment]
I'm curious whether the [ArchUnit test still fails](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=56031&view=logs&j=fc5181b0-e452-5c8f-68de-1097947f6483&t=995c650b-6573-581c-9ce6-7ad4cc038461&l=22460). I'm not able to reproduce it locally:
```bash
./mvnw -pl flink-architecture-tests/flink-architecture-tests-production surefire:test
```
Do you have any experiences with running the archunit tests locally?

**Comment by 1996fanrui** [comment]
> I'm curious whether the [ArchUnit test still fails](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=56031&view=logs&j=fc5181b0-e452-5c8f-68de-1097947f6483&t=995c650b-6573-581c-9ce6-7ad4cc038461&l=22460). I'm not able to reproduce it locally:
> 
> ```shell
> ./mvnw -pl flink-architecture-tests/flink-architecture-tests-production surefire:test
> ```
> 
> Do you have any experiences with running the archunit tests locally?

Sorry for that, I didn't run it locally before. I'm not sure whether re-run is fine.

**Review by 1996fanrui — APPROVED** [review]
Thanks @XComp for fixing the CI, LGTM.

**Pull request merged**